src/AspNetCore/Src/aspnetcoreapplication.cxx

```diff
@@ -0,0 +1,516 @@
+    _In_ VOID* pvRequstHandlerContext,
+    _In_ VOID* pvShutdownHandlerContext
+)
+{
+    ASPNETCORE_APPLICATION::GetInstance()->SetCallbackHandles(request_handler, shutdown_handler, pvRequstHandlerContext, pvShutdownHandlerContext);
+}
+
+extern "C" __declspec(dllexport)
+HTTP_REQUEST*
+http_get_raw_request(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->GetRequest()->GetRawHttpRequest();
+}
+
+extern "C" __declspec(dllexport)
+HTTP_RESPONSE*
+http_get_raw_response(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->GetResponse()->GetRawHttpResponse();
+}
+
+extern "C" __declspec(dllexport) VOID http_set_response_status_code(
+    _In_ IHttpContext* pHttpContext,
+    _In_ USHORT statusCode,
+    _In_ PCSTR pszReason
+)
+{
+    pHttpContext->GetResponse()->SetStatus(statusCode, pszReason);
+}
+
+extern "C" __declspec(dllexport)
+HRESULT
+http_post_completion(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->PostCompletion(0);
+}
+
+extern "C" __declspec(dllexport)
+VOID
+http_indicate_completion(
+    _In_ IHttpContext* pHttpContext,
+    _In_ REQUEST_NOTIFICATION_STATUS notificationStatus
+)
+{
+    pHttpContext->IndicateCompletion(notificationStatus);
+}
+
+extern "C" __declspec(dllexport)
+VOID
+http_get_completion_info(
+    _In_ IHttpCompletionInfo2* info,
+    _Out_ DWORD* cbBytes,
+    _Out_ HRESULT* hr
+)
+{
+    *cbBytes = info->GetCompletionBytes();
+    *hr = info->GetCompletionStatus();
+}
+
+extern "C" __declspec(dllexport)
+BSTR // TODO probably should make this a wide string
+http_get_application_full_path()
+{
+    return SysAllocString(ASPNETCORE_APPLICATION::GetInstance()->GetConfig()->QueryApplicationFullPath()->QueryStr());
+}
+
+extern "C" __declspec(dllexport)
+HRESULT
+http_read_request_bytes(
+    _In_ IHttpContext* pHttpContext,
+    _In_ CHAR* pvBuffer,
+    _In_ DWORD cbBuffer,
+    _In_ PFN_ASYNC_COMPLETION pfnCompletionCallback,
+    _In_ VOID* pvCompletionContext,
+    _In_ DWORD* pDwBytesReceived,
+    _In_ BOOL* pfCompletionPending
+)
+{
+    auto pHttpRequest = (IHttpRequest3*)pHttpContext->GetRequest();
+
+    BOOL fAsync = TRUE;
+
+    HRESULT hr = pHttpRequest->ReadEntityBody(
+        pvBuffer,
+        cbBuffer,
+        fAsync,
+        pfnCompletionCallback,
+        pvCompletionContext,
+        pDwBytesReceived,
+        pfCompletionPending);
+
+    if (hr == HRESULT_FROM_WIN32(ERROR_HANDLE_EOF))
+    {
+        // We reached the end of the data
+        hr = S_OK;
+    }
+
+    return hr;
+}
+
+extern "C" __declspec(dllexport)
+HRESULT
+http_write_response_bytes(
+    _In_ IHttpContext* pHttpContext,
+    _In_ HTTP_DATA_CHUNK* pDataChunks,
+    _In_ DWORD nChunks,
+    _In_ PFN_ASYNC_COMPLETION pfnCompletionCallback,
+    _In_ VOID* pvCompletionContext,
+    _In_ BOOL* pfCompletionExpected
+)
+{
+    auto pHttpResponse = (IHttpResponse2*)pHttpContext->GetResponse();
+
+    BOOL fAsync = TRUE;
+    BOOL fMoreData = TRUE;
```

Comment:
why do you hardcode those values

Reply:
@davidfowl Can we just assume that async and more data will always be TRUE?